Feat training

furyhawk opened the pull request:
Add feature for Training pipeline for image classification.

sabrimansor committed:
refactor: improve implementation for printing model summary

sabrimansor committed:
Merge remote-tracking branch 'refs/remotes/origin/feat-training' into feat-training

sabrimansor committed:
feat: added implementation for tensorflow trainer model summary

furyhawk committed:
refact: clean up for CICD

furyhawk committed:
refact: remove dev container

furyhawk committed:
refact: typing checks

sabrimansor committed:
feat: initial structure for training pipeline documentation

furyhawk committed:
fix: revert to Union

furyhawk committed:
fix: Dict class instead of dict object

sabrimansor committed:
feat: init commit for different use case custom parameters

sabrimansor committed:
Merge remote-tracking branch 'refs/remotes/origin/feat-training' into feat-training

furyhawk committed:
refact: add typing to data_adapter

furyhawk committed:
fix: remove unused function

furyhawk committed:
refact: remove AbstractDataAdapter, dataclass DataAdapter

sabrimansor committed:
feat: added framework tab for custom parameters doc page

sabrimansor committed:
feat: added manual seed during training setup

sabrimansor committed:
Merge remote-tracking branch 'refs/remotes/origin/feat-training' into feat-training

furyhawk committed:
fix: remove print, revert return type to python 3.8

furyhawk committed:
Merge remote-tracking branch 'upstream/dev' into feat-training

furyhawk committed:
doc: black format

sabrimansor committed:
feat: setting timm as default pytorch model adapter

furyhawk committed:
fix: typing check

sabrimansor committed:
feat: added initial config table for callbacks and metrics section

sabrimansor committed:
Merge remote-tracking branch 'refs/remotes/origin/feat-training' into feat-training

furyhawk committed:
refact: typing check

furyhawk committed:
refact: type check

sabrimansor committed:
feat: added initial config table for store, trainer and model analysis

sabrimansor committed:
Merge remote-tracking branch 'refs/remotes/origin/feat-training' into feat-training

furyhawk committed:
refact: typing checks

furyhawk committed:
refact: add type checking hints

furyhawk committed:
fix: type reassignment error

furyhawk committed:
refact: remaining type hints

furyhawk committed:
fix: self.metrics_tensor rename back to self.metrics, and training metrics assignment error

frankcaoyun committed:
fix: added tensorflow packages in the requirements file for mac.

sabrimansor committed:
feat: added config from trainer model and data modules

sabrimansor committed:
refactor: call model_artifacts_dir directly from trainer object

sabrimansor committed:
Merge remote-tracking branch 'refs/remotes/origin/feat-training' into feat-training

furyhawk committed:
doc: document data adapter

furyhawk committed:
doc: dataset, transformation documentation

furyhawk committed:
doc: Using custom dataset

furyhawk committed:
fix: remove subscription to support python v3.8

sabrimansor committed:
fix: standardised styling for key column

sabrimansor committed:
Merge remote-tracking branch 'refs/remotes/origin/feat-training' into feat-training

sabrimansor committed:
refactor: added some changes

frankcaoyun committed:
fix: update requirement pytorch and tensorflow version to be compatible within the same venv.

frankcaoyun committed:
feat: added two basic training testing scripts for pytorch and tensorflow to test if CUDA is working.

sabrimansor committed:
feat: getting started initial commit

sabrimansor committed:
Merge remote-tracking branch 'refs/remotes/origin/feat-training' into feat-training

sabrimansor committed:
Refactor: remove global_train_params reference

furyhawk committed:
fix: lint and missing copyright or license headers

furyhawk committed:
fix: use Typing subscription

furyhawk committed:
fix: missing type return

gao-hongnan commented on peekingduck/training/src/data/data_module.py:
```diff
@@ -0,0 +1,196 @@
+    def __init__(
+        self,
+        cfg: DictConfig,
+        **kwargs: Dict[str, Any],
```
`kwargs` seems to be unused in the class, consider either adding `self.kwargs = kwargs` or updating via `self.__dict__.update(kwargs)`.

gao-hongnan commented on peekingduck/training/src/data/data_adapter.py:
```diff
@@ -0,0 +1,99 @@
+class DataAdapter:
+    """Adapter for different framework"""
+
+    cfg: DictConfig = field()
```
Is there a reason to use `field()` instead of just type hinting as `cfg: DictConfig`?

I think some usages of `field()` is when your attribute is mutable, setting default values and controlling serialization in `__repr__`.

gao-hongnan commented on peekingduck/training/src/data/data_adapter.py:
```diff
@@ -0,0 +1,99 @@
+    def train_dataloader(
+        self, dataset: AbstractDataSet, transforms: Compose
+    ) -> Optional[Any]:
+        assert self.cfg.adapter_type in [
+            "pytorch",
+            "tensorflow",
+        ], f"Unsupported framework {self.cfg.adapter_type}"
```
For the assert checks in `train_dataloader`, `validation_dataloader` and `test_dataloader`, are we able to move them to `__post_init__` so that it is checked only once upon construction?

gao-hongnan commented on peekingduck/training/src/data/dataset.py:
```diff
@@ -0,0 +1,224 @@
+import cv2
+import pandas as pd
+
+from .. import config
```
Use absolute imports as the convention.

gao-hongnan commented on peekingduck/training/src/data/dataset.py:
```diff
@@ -0,0 +1,224 @@
+        self.num_channels = num_channels
+        self.shuffle = shuffle
+
+        self.list_IDs = df[x_col].values
```
Consider following the snake case convention (which applies to the rest of the repo).

gao-hongnan commented on peekingduck/training/src/data/dataset.py:
```diff
@@ -0,0 +1,224 @@
+    def _on_epoch_end(self) -> None:
+        "Updates indexes after each epoch"
+        self.indexes = np.arange(len(self.list_IDs))
+        if self.shuffle == True:
```
Consider using `if self.shuffle` or `self.shuffle is True`. However, as the reference below mentioned, be mindful of using it with `numpy` and its derivatives.

See [reference on singleton-comparison](https://stackoverflow.com/questions/27276610/boolean-identity-true-vs-is-true).

gao-hongnan commented on peekingduck/training/src/data/data_module.py:
```diff
@@ -0,0 +1,196 @@
+        )
+        self.dataset_loader: Union[DataAdapter, None] = None  # Setup in self.setup()
+
+    def get_train_dataset(self) -> DataAdapter:
```
Consider renaming `get_train_dataset` to `get_train_dataloader` since we are returning a `DataLoader` object. Same for `get_validation_dataset`.

furyhawk replied:
Strange, I got `get_test_dataloader` already there. Note to future reader: Tensorflow dataloader is actually a dataset(`tf.keras.utils.Sequence`).

gao-hongnan replied:
Noted on that, I realised that afterwards, can discuss further on this naming convention since in `DataAdapter`, there is mention of `self.loader`.

gao-hongnan commented on peekingduck/training/src/model/pytorch_model.py:
```diff
@@ -0,0 +1,117 @@
+            # create and reset the classifier layer
+            head = self.create_head(in_features)
+            rsetattr(self.backbone, last_layer_name, head)
+            model = self.backbone
+
+        elif self.adapter == "timm":
+            self.backbone.reset_classifier(num_classes=self.model_config.num_classes)  # type: ignore
+            model = self.backbone
```
Consider moving `model = self.backbone` outside the `if-else` block since they are common in both the `if-elif` block.

furyhawk committed:
fix: PR #747 data_module scripts fixes

sabrimansor committed:
feat: update to trainer documentation

sabrimansor committed:
Merge remote-tracking branch 'refs/remotes/origin/feat-training' into feat-training

sabrimansor committed:
feat: implement separate learning rate for pytorch fine tuning

furyhawk committed:
fix: black formatter

sabrimansor committed:
feat: updated trainer class config description

sabrimansor committed:
Merge remote-tracking branch 'refs/remotes/origin/feat-training' into feat-training

furyhawk committed:
fix: type string

frankcaoyun committed:
fix: 1. link the use_case with the main config file; 2. update the index link for pytorch 3. remove printing cudnn version in the pytorch test file (will give error when having both tensorflow and pytorch together but doesn't affect training.)

furyhawk committed:
fix: resolve PR #747 get_train_dataset to get_train_dataloader

gao-hongnan commented on peekingduck/training/src/model/pytorch_model.py:
```diff
@@ -0,0 +1,116 @@
+            raise ValueError(f"Adapter {self.adapter} not supported.")
+
+        # freeze the backbone because it was trainable by default
+        freeze_all_params(backbone)
```
Currently, the `fine_tune_modules` is not used, will that be a feature in future?

gao-hongnan commented on peekingduck/training/src/model/pytorch_model.py:
```diff
@@ -0,0 +1,116 @@
+
+        return backbone
+
+    def create_head(self, in_features: int) -> nn.Module:
```
This seems to violate the Liskov Substitution Principle as the base class's method `create_head` takes in no arguments while the concrete implementation does.

See [Liskov Substitution Principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle) (the L in SOLID) for details on this potential issue.

gao-hongnan commented on peekingduck/training/src/model/tensorflow_model.py:
```diff
@@ -0,0 +1,52 @@
+    """Generic TensorFlow image classification model."""
+
+    @classmethod
+    def create_model(cls, model_cfg: DictConfig) -> tf.keras.Model:
```
Same issue as the `create_head`, consider to have the same set of arguments in the base class as well.

gao-hongnan commented on peekingduck/training/src/model/tensorflow_model.py:
```diff
@@ -0,0 +1,52 @@
+        num_classes = model_cfg.num_classes
+        prediction_layer_name = "prediction_modified"
+
+        model: tf.keras.Model = getattr(tf.keras.applications, model_name)(
+            include_top=True, weights="imagenet"
+        )
+
+        # exclude the existing prediction layer
+        x = model.layers[-2].output
+        # create the new prediction layer
+        predictions = tf.keras.layers.Dense(
+            num_classes, activation="softmax", name=prediction_layer_name
+        )(x)
```
Consider shifting configurable parameters to hydra's config management system. As an hypothetical example, some may prefer to use `activation="sigmoid"` for binary classification with 1 output neuron instead of 2.

gao-hongnan commented on peekingduck/training/src/callbacks/base.py:
```diff
@@ -0,0 +1,160 @@
+    """
+    cb_list = []
+    for cb in callbacks:
+        if type(cb) is DictConfig:
```
Consider using `isinstance` for type checking.

The reason is subclass can be also checked, so it is a stronger condition than using `type` equality check.

If justified, consider making changes to other parts where `type` checking is used.

See [here](https://stackoverflow.com/questions/30049410/why-use-isinstance-instead-of-type).

gao-hongnan commented on peekingduck/training/src/metrics/pytorch_metrics.py:
```diff
@@ -0,0 +1,79 @@
+                )
+
+        else:
+            raise TypeError
```
It may be good to include more detailed error/exception messages.

gao-hongnan commented on peekingduck/training/src/metrics/tensorflow_metrics.py:
```diff
@@ -0,0 +1,43 @@
+
+class TensorflowMetrics:
+    def get_metric(
+        self, metric_name: str, parameters: Union[DictConfig, dict] = {}
```
Using dictionary default can cause behaviours mentioned in [here](https://stackoverflow.com/questions/26320899/why-is-the-empty-dictionary-a-dangerous-default-value-in-python).

In general, mutable defaults can cause mutation downstream (was a first hand victim before).

ongtw replied:
Does `parameters` need a default value? If not, it should be specified such that the caller always has to pass in something.

frankcaoyun committed:
feat: add in config for fine-tuning the whole model.

furyhawk committed:
fix: PR #747 fix mutable defaults, `isinstance` type checking,

sabrimansor committed:
fix: wrongly referenced tensorflow earlystopping monitored metrics

frankcaoyun committed:
Merge pull request #6 from frankcaoyun/feat-training-model

feat: add in config for fine-tuning the whole model.

furyhawk committed:
fix: default `fine_tune_all`: False

furyhawk committed:
fix: PR #747 pylint errors

frankcaoyun committed:
feat: add in installation guide for Linux and Windows

sabrimansor committed:
fix: tensorflow print train summary error

sabrimansor committed:
Merge remote-tracking branch 'refs/remotes/origin/feat-training' into feat-training

frankcaoyun committed:
feat: added in wandb setup and test run in the getting started section.

sabrimansor committed:
feat: updated getting started renamed to overview

sabrimansor committed:
Merge remote-tracking branch 'refs/remotes/origin/feat-training' into feat-training

sabrimansor committed:
feat: display config directory in overview

sabrimansor committed:
feat: added config directory tree and reference each section to its relative config file

sabrimansor committed:
fix: added backslash to config folder on config directory tree

sabrimansor committed:
fix: updated installation docs for mac

sabrimansor committed:
refactor: replace code block line icons with line numbering

sabrimansor committed:
feat: updated miniforge link to install for mac installation

frankcaoyun committed:
fix: clean up wandb and test run sections.

furyhawk committed:
fix: train_summary for tensorflow

frankcaoyun committed:
feat: add in description table for overview.

sabrimansor committed:
fix: updated installation command for masOS

frankcaoyun committed:
feat: add in command line instructions in the "configuring training parameters" section.

sabrimansor committed:
refactor: update daigram to make it look more compact

sabrimansor committed:
Merge remote-tracking branch 'refs/remotes/origin/feat-training' into feat-training

frankcaoyun committed:
feat: Update pytorch model config description in docs.

frankcaoyun committed:
Merge branch 'feat-training' of https://github.com/furyhawk/PeekingDuck into feat-training

sabrimansor committed:
update tensorflow progress bar to use tqdm instead of the keras progbar

sabrimansor committed:
Merge remote-tracking branch 'refs/remotes/origin/feat-training' into feat-training

frankcaoyun committed:
fix: update tensorflow progress bar and update messages for unfreezing and fine-tuning.

frankcaoyun committed:
Merge branch 'feat-training' of https://github.com/furyhawk/PeekingDuck into feat-training

frankcaoyun committed:
feat: Update docs for tensorflow model. Fix table cell width issues when using a smaller browser size (credits to Sabri!)

furyhawk committed:
fix: resolve self.callbacks check if empty

furyhawk committed:
fix: pylint: Your code has been rated at 9.08/10

sabrimansor committed:
fix: added fix for black error

sabrimansor committed:
feat: added config file label for the path to the config file

sabrimansor committed:
feat: added todo messages for later easier to change

sabrimansor committed:
feat: added document convention section

frankcaoyun committed:
fix: clean up the pytorch model adapter description and move to FAQ.

sabrimansor committed:
feat: added grey styling for configs that should not be changed

sabrimansor committed:
refactor: moved the documentation convention section to the usecase page instead

sabrimansor committed:
refactor: swap the template and sample

sabrimansor committed:
Merge remote-tracking branch 'refs/remotes/origin/feat-training' into feat-training

sabrimansor committed:
feat: grey styling css

furyhawk committed:
fix: add !data/**/*.csv exception

furyhawk committed:
fix: remove *.csv from .gitignore

furyhawk committed:
feat: training csv for datasets

sabrimansor committed:
fix: removed target user heading

furyhawk committed:
Merge branch 'feat-training' into feat-training

furyhawk committed:
fix: clear callbacks pylint error

furyhawk committed:
fix: training-pipeline requirement

furyhawk committed:
fix: .pylintrc for module import

furyhawk committed:
fix: resolve Missing module docstring

frankcaoyun committed:
fix: 1. Update requirements.txt with specific version of torchvision
2. Add in separate epochs setting for fine-tuninng
3. Refactor fine_tuning logic for pytorch trainer
3. Add in value checking for pytorch fine_tune setting. Perform it before strat training in PT and TF trainer

frankcaoyun committed:
Merge branch 'feat-training' of https://github.com/furyhawk/PeekingDuck into feat-training

furyhawk committed:
fix: resolve data,model,trainer pylint issues

furyhawk committed:
fix: add missing __init__.py files

sabrimansor committed:
fix: black formatting error

furyhawk committed:
fix: type hinting Union[nn.Module, Callable]

furyhawk committed:
fix: remove re-decaration

furyhawk committed:
fix: typeguard new check_type

furyhawk committed:
fix: typeguard == 2.13.3

furyhawk committed:
test: configs test

furyhawk committed:
fix: space

furyhawk committed:
feat: check for null `model_artifacts_dir`

furyhawk committed:
test: training-pipeline test

furyhawk committed:
fix: black formatter

furyhawk committed:
fix: tensorflow metric AUC `from_logits` deprecated

furyhawk committed:
fix: remove `num_labels` to fit tensorflow2.2.0

furyhawk committed:
fix: tensorflow2.2 remove `expand_nested=True`

furyhawk committed:
test: remove tensorflow downgrade linter test

furyhawk committed:
feat: return training history

furyhawk committed:
test: test return training history validation loss

furyhawk committed:
feat: add `tensorflow-intel` to valid `tensorflow` package list

furyhawk committed:
test: requirements_lint.txt to requirements.txt

sabrimansor committed:
fix: added missing packages

sabrimansor committed:
Merge remote-tracking branch 'refs/remotes/origin/feat-training' into feat-training

sabrimansor committed:
fix: dependency conflict between sphinx packages and docutils

sabrimansor committed:
fix: for h5py conflict and sphinx package conflict

furyhawk committed:
fix: remove `imagesize` and `tabulate` packages.

furyhawk committed:
fix: replace deprecated `GabrielBB/xvfb-action@v1.6` with `coactions/setup-xvfb@v1`

furyhawk committed:
fix: pytest for linux platform

furyhawk committed:
test: create keras example test

sabrimansor committed:
feat: amended the docs removing typos and incorrect sentence structures

sabrimansor committed:
Merge remote-tracking branch 'refs/remotes/origin/feat-training' into feat-training

furyhawk committed:
test: unit test for Tensorflow datasource

furyhawk committed:
refactor: better naming of test functions, tuple to typing Tuple

furyhawk committed:
fix: add `tensorflow-datasets` for cicd unit test

furyhawk committed:
refactor: rename valid to validation

furyhawk committed:
build: constraint albumentations ~= 1.3

furyhawk committed:
test: configurations detailed

furyhawk committed:
test: add tensorflow dataloader unit test

ongtw commented on docs/source/training_pipeline/01_getting_started.rst:
```diff
@@ -0,0 +1,279 @@
+.. raw:: html
+
+   <h6>TensorFlow and PyTorch Models</h6>
+Supports pre-trained models from both PyTorch and TensorFlow. The selection of the model can be customized in the configuration files.
```
For `Features` section, need to standardize on the sentence structure style. Above point uses clauses (partial sentences), e.g. "Supports pre-trained models ...", but `Analyze Runs` point below uses full sentences.
As this is _not_ a marketing document, we should write in complete sentences.

frankcaoyun replied:
Well noted. Will update to the full sentences.

frankcaoyun committed:
fix: Update the overview page for training pipeline for better clarity.

ongtw commented on docs/source/training_pipeline/index.rst:
```diff
@@ -0,0 +1,15 @@
+.. _training_pipeline:
+
+***********************
+Training your own model
```
Wrong stylization compared with the other Tutorials titles, see image attached:
![image](https://user-images.githubusercontent.com/13075447/229012912-7af0f1c3-2f13-432d-b598-9760b04baba0.png)

frankcaoyun replied:
Will update the words as capitalized.

ongtw commented on docs/source/training_pipeline/01_getting_started.rst:
```diff
@@ -0,0 +1,279 @@
+.. raw:: html
+
+   <h6>Save the trained model</h6>
+The trained models will be saved in the output folder.
```
This sounds weird as a feature: the trained model will certainly have to be saved somewhere, so "saving" is a given. Would be weird to have a product that trains without saving, right?

frankcaoyun replied:
Will remove and mention in the model analysis (analyze the saved model).

ongtw commented on docs/source/training_pipeline/01_getting_started.rst:
```diff
@@ -0,0 +1,279 @@
+         conda activate pkd-training
+
+         # Clone PeekingDuck repository
+         git clone <TODO: PeekdingDuck repository>
```
I think we can replace the `<TODO: ...>` with the actual PeekingDuck GitHub repo URL.

sabrimansor replied:
replaced with peekingduck repo

ongtw commented on docs/source/training_pipeline/01_getting_started.rst:
```diff
@@ -0,0 +1,279 @@
+         && mkdir -p $CONDA_PREFIX/etc/conda/deactivate.d \
+         && echo 'export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$CONDA_PREFIX/lib/'>\
+         $CONDA_PREFIX/etc/conda/activate.d/env_vars.sh && \
+         echo 'unset LD_LIBRARY_PATH'>\
```
Above script of 5 lines is doing a lot of things and should be split up so that each line does a single task.
Essentially, it's doing:
1. Check for existence of `activate.d` / `deactivate.d` and create it if not found
2. Add the LD_LIBRARY_PATH into `env_vars.sh`
3. Print a message

Lumping everything into one makes it hard to read and maintain.

frankcaoyun replied:
Noted. The initial consideration was for the user to execute the commands easily with a single command, but agree with the readability. Will separate them out.

ongtw commented on docs/source/training_pipeline/01_getting_started.rst:
```diff
@@ -0,0 +1,283 @@
+   # Login to weights & biases
+   wandb login
+
+Weights & Biases comes out of the box with the training pipeline but in the case where you do not want to use weights & biases you can disable it as such.
```
This is incorrect phrasing: we are not providing W&B "out of the box" with the training pipeline.
What you mean is the training pipeline is designed to work with W&B by default, but in case the user does not want to use W&B, it can be disabled.
The above should express the intention correctly as is.

sabrimansor replied:
these lines have been removed.

ongtw commented on docs/source/training_pipeline/01_getting_started.rst:
```diff
@@ -0,0 +1,283 @@
+   pip install wandb
+
+   # Login to weights & biases
+   wandb login
```
For above `code-block` (and all other related portions below and elsewhere), our convention for showing screen input/output is to use `.. admonition:: Terminal Session` together with the actual color coded display. You can reference the sample in https://raw.githubusercontent.com/aisingapore/PeekingDuck/main/docs/source/tutorials/01_hello_cv.rst

The terminal display should so both the user input and the expected output.
And to be instructive and complete, you will want to show the prompt where the user inputs the PAT into W&B, because this is the last step.

ongtw commented on docs/source/training_pipeline/01_getting_started.rst:
```diff
@@ -0,0 +1,283 @@
+   # Pytorch
+   python ./peekingduck/training/main.py debug=True framework=pytorch
+
+View the results of each run at the specified output folder directory: :mod:`\./Peekingduck/outputs/\<PROJECT_NAME\>/\<DATE_TIME\>`
```
The `<PROJECT_NAME>` is used here but not referenced / introduced anywhere up above.
This creates confusion as the reader will wonder what is this `<PROJECT_NAME>`.
Need to rewrite above to make sure the flow is coherent and the information is complete. If need to use project name, then specify and set it before arriving at this line.

ongtw commented on docs/source/training_pipeline/01_getting_started.rst:
```diff
@@ -0,0 +1,283 @@
+
+View the results of each run at the specified output folder directory: :mod:`\./Peekingduck/outputs/\<PROJECT_NAME\>/\<DATE_TIME\>`
+
+After running either of the commands without any errors, read on to the next page to see how you can configure the parameters for a more comprehensive training.
```
I think can remove this line: first part sounds strange (are we expecting the user to encounter and fix errors?) and there is no need to say the obvious. If anything, the intro in for the next section should specify that "This section describes the training parameters and their possible configuration, etc..."

sabrimansor replied:
these lines have been removed and rephrased.

ongtw commented on docs/source/training_pipeline/02_configuring_training_parameters/index.rst:
```diff
@@ -0,0 +1,143 @@
+
+This section describes how you can configure the pipeline to suite your model training.
+This guide is written with the assumption that the user has prior knowledge of the technical terms associated with typical Machine Learning training pipelines.
+Readers will also need to have prior knowledge of the python programming language and yaml configuration files. 
```
Don't need above line as both knowledge are already assumed when using PeekingDuck.

sabrimansor replied:
these lines have been removed.

ongtw reviewed (changes requested):
Some preliminary comments on the docs.

sabrimansor committed:
fix: toctree recursive error and reviewed section 3 documentations

sabrimansor committed:
Merge remote-tracking branch 'refs/remotes/origin/feat-training' into feat-training

sabrimansor committed:
fix: corrected some sentence structure

furyhawk committed:
test: force device to cpu

furyhawk committed:
test: add trainer unit test

furyhawk committed:
cicd: disable Weights & Biases

frankcaoyun committed:
fix: Update getting started page contents and formats based on the feedbacks.

furyhawk committed:
test: unset device=cpu

furyhawk committed:
test: set device=cpu

furyhawk committed:
test: disable weight and biases

furyhawk committed:
test: unit test for train_loader and validation_loader

furyhawk committed:
test: enable weight & biases, add pytorch dataloader test

furyhawk committed:
refactor: correct type hinting for AbstractDataSet

furyhawk committed:
test: test tensorflow trainer only

furyhawk committed:
test: add tensorflow model test

furyhawk committed:
test: 10 classes fix, val_sparse_categorical_accuracy metric

furyhawk committed:
test: replace loss object

furyhawk committed:
refactor: rename model_cfg to model_config

furyhawk committed:
test: add test_tensorflow_model feature extraction test

furyhawk committed:
test: temp disable failed test

furyhawk committed:
refactor: type hint tf.keras.Model

furyhawk committed:
feat: input_tensor for custom image shape

furyhawk committed:
test: enable tensorflow trainer test

furyhawk committed:
test: enable tensorflow training_pipeline test

furyhawk committed:
test: set metric to SparseCategoricalCrossentropy

furyhawk committed:
test: enable tensorflow training pipeline test for windows platform

furyhawk committed:
test: test VGG16 in linux only

furyhawk committed:
test: test_tensorflow_model_with_loss expected to 0.4

frankcaoyun committed:
docs: Update the getting started page for:
1. Format consistency
2. Cross-reference

frankcaoyun committed:
Merge branch 'feat-training' of https://github.com/furyhawk/PeekingDuck into feat-training

frankcaoyun committed:
docs: Update the description for configuring training parameters.

furyhawk committed:
test: increase training data to 2% for less variance

frankcaoyun committed:
docs: Reformat the overview folder structure following the peekingduck style.

frankcaoyun committed:
Merge branch 'feat-training' of https://github.com/furyhawk/PeekingDuck into feat-training

furyhawk committed:
doc: dataset download link

frankcaoyun committed:
docs:
1. Update the folder structure for consistent formats.
2. Update descriptions of the yaml file preparation for better clarity.

sabrimansor committed:
feat: added tensorflow ModelCheckpoint callback to default callback list

sabrimansor committed:
fix: specifying tensorflow-dataset library to  version 4.8.3

furyhawk committed:
fix: resolve tensorflow-datasets == 4.9.0 error, TIL bug fix

furyhawk committed:
Merge branch 'feat-training' of https://github.com/furyhawk/PeekingDuck into feat-training

ongtw commented on data/cifar10/train.csv:
`data/cifar10` is a huge 237MB folder.
Can we look at not including `data/cifar10` and getting the user to download this on demand instead?

Of course, any files that we need internally for use with our training pipeline should be included in this (empty) folder.

furyhawk replied:
Resolved by removing the dataset from the repository. `cifar10` is uploaded to https://storage.googleapis.com/peekingduck/data/cifar10.zip

ongtw commented on docs/source/_static/css/pkdk.css:
```diff
@@ -179,6 +179,10 @@ div.highlight span.pkd-raw-terminal:before {
 }
 
+.grey {
+   color: #c2c2c2;
```
Isn't `grey` a predefined CSS color? Why need to use `#c2c2c2` RGB?

frankcaoyun replied:
updated as grey

ongtw commented on docs/source/training_pipeline/01_getting_started.rst:
```diff
@@ -0,0 +1,260 @@
+Overview
+********
+
+* :ref:`getting-started-test-run`
```
This looks weird and out-of-place by itself:
![image](https://user-images.githubusercontent.com/13075447/232669640-34110c8f-bd74-4371-a9fb-a79afef395cf.png)
We should remove this.

frankcaoyun replied:
Removed.

ongtw commented on docs/source/faq.rst:
```diff
@@ -45,4 +45,11 @@ Why does the output screen flash briefly and disappear on my second run?
\ No newline at end of file
+`GitHub issue <https://github.com/aisingapore/PeekingDuck/issues/630>`_ for more details.
+
+How to choose the pretrained model library between ``timm`` and ``torchvision`` for PyTorch adapter?
```
The English in this FAQ reads weird. Do you mean:
`How to choose between 'timm' and 'torchvision' model libraries for the PyTorch adapter?`

frankcaoyun replied:
Noted and updated.

ongtw commented on docs/source/training_pipeline/01_getting_started.rst:
```diff
@@ -0,0 +1,260 @@
+==============
+
+Training a custom model with custom datasets is one of the key tasks when building computer vision solutions.
+
```
Remove this blank line to create a paragraph of two sentences, to reduce introduction length.

frankcaoyun replied:
Noted.

ongtw commented on docs/source/training_pipeline/01_getting_started.rst:
```diff
@@ -0,0 +1,260 @@
+
+Peeking Duck allows you to train a custom computer vision model with your own dataset without the need to write the boilerplate codes.
+
+This section will introduce you the key features of the training pipeline and how to get started.
```
"This section will introduce you `to` the ..."

sabrimansor replied:
Resolved in commit 83661dbc12df0138d3abc1256ad7aed2b7540732

ongtw commented on docs/source/training_pipeline/01_getting_started.rst:
```diff
@@ -0,0 +1,260 @@
+Peeking Duck allows you to train a custom computer vision model with your own dataset without the need to write the boilerplate codes.
+
+This section will introduce you the key features of the training pipeline and how to get started.
+The following sections will introduce you how to train your own model with your own data by customizing the training parameters.
```
"The following sections `describes` how to ..."

sabrimansor replied:
Resolved in commit 83661dbc12df0138d3abc1256ad7aed2b7540732

ongtw commented on docs/source/training_pipeline/01_getting_started.rst:
```diff
@@ -0,0 +1,260 @@
+This section will introduce you the key features of the training pipeline and how to get started.
+The following sections will introduce you how to train your own model with your own data by customizing the training parameters.
+
+The current version of PeekingDuck supports image classification model training. Object detection are being developed and will be available in future releases.
```
"Currently, PeekingDuck support model training for image classification."
"Model training for object detection will be available in a future release."

sabrimansor replied:
Resolved in commit 83661dbc12df0138d3abc1256ad7aed2b7540732

ongtw commented on docs/source/training_pipeline/01_getting_started.rst:
```diff
@@ -0,0 +1,260 @@
+
+   <h6>Analyze right after training</h6>
+
+PeekingDuck uses Weights & Biases to analyze and visualize the training process and performances of the saved models. Refer to the :ref:`setting-up-weights-and-biases` section below for more details.
```
Lines 26-48: Do not use `raw:: html` as the rendering looks unimpressive:
![image](https://user-images.githubusercontent.com/13075447/232671086-bbdb29d9-2539-4830-9c34-f33d55b343a4.png)
The `<h6>` font and the `How the training pipeline works` fonts are identical and reader will mistake the last one for a "feature".
Just use normal, simple Sphinx bullets **_without html embellishments_**, like `01_hello_cv.rst` lines 215 and 221.

sabrimansor replied:
Resolved in commit 83661dbc12df0138d3abc1256ad7aed2b7540732

ongtw commented on docs/source/training_pipeline/01_getting_started.rst:
```diff
@@ -0,0 +1,260 @@
+PeekingDuck uses Weights & Biases to analyze and visualize the training process and performances of the saved models. Refer to the :ref:`setting-up-weights-and-biases` section below for more details.
+
+
+How the training pipeline works
```
Stylize: `How the Training Pipeline Works`

sabrimansor replied:
Resolved in commit 83661dbc12df0138d3abc1256ad7aed2b7540732

ongtw commented on docs/source/training_pipeline/01_getting_started.rst:
```diff
@@ -0,0 +1,260 @@
+How the training pipeline works
+===============================
+
+PeekingDuck training pipeline is designed with cross-platform compatibility in mind. A high-level overview of the architecture is shown here:
```
"PeekingDuck's training pipeline is designed to be compatible with PyTorch and Tensorflow. The architecture diagram is shown below:"

sabrimansor replied:
Resolved in commit 83661dbc12df0138d3abc1256ad7aed2b7540732

ongtw commented on docs/source/training_pipeline/01_getting_started.rst:
```diff
@@ -0,0 +1,260 @@
+
+#. Organize your data with the required format 
+#. Place them at the designated directory
+#. Personalise your training if necessary 
```
"Customize training parameters, if necessary."

sabrimansor replied:
Resolved in commit 83661dbc12df0138d3abc1256ad7aed2b7540732

ongtw commented on docs/source/training_pipeline/01_getting_started.rst:
```diff
@@ -0,0 +1,260 @@
+The general workflow when using the training pipeline is:
+
+#. Organize your data with the required format 
+#. Place them at the designated directory
```
Combine these two into one sentence:
"Organize training data with the required format in the designated directory."

sabrimansor replied:
Resolved in commit 83661dbc12df0138d3abc1256ad7aed2b7540732

ongtw commented on docs/source/training_pipeline/01_getting_started.rst:
```diff
@@ -0,0 +1,260 @@
+#. Organize your data with the required format 
+#. Place them at the designated directory
+#. Personalise your training if necessary 
+#. Execute from the terminal
```
"Start model training from the terminal."

sabrimansor replied:
Resolved in commit 83661dbc12df0138d3abc1256ad7aed2b7540732

ongtw commented on docs/source/training_pipeline/01_getting_started.rst:
```diff
@@ -0,0 +1,260 @@
+#. Personalise your training if necessary 
+#. Execute from the terminal
+
+Refer to the next section to get started.
```
Remove this redundant sentence.

sabrimansor replied:
Resolved in commit 83661dbc12df0138d3abc1256ad7aed2b7540732

ongtw commented on docs/source/training_pipeline/01_getting_started.rst:
```diff
@@ -0,0 +1,260 @@
+ Getting Started
+================
+
+This guide explains how you can prepare your environment and test out the default pipeline using either TensorFlow or Pytorch.
```
Stylize `PyTorch`

sabrimansor replied:
Resolved in commit 83661dbc12df0138d3abc1256ad7aed2b7540732

ongtw commented on docs/source/training_pipeline/01_getting_started.rst:
```diff
@@ -0,0 +1,260 @@
+
+This guide explains how you can prepare your environment and test out the default pipeline using either TensorFlow or Pytorch.
+
+It is highly recommended to create a new conda environment as instructed below. Installing PeekingDuck dependencies into existing conda environments may cause unexpected behaviours.
```
"It is highly recommended to create a new conda environment as instructed below, to avoid package conflicts and unexpected behaviors when running from an existing conda environment.

sabrimansor replied:
Resolved in commit 83661dbc12df0138d3abc1256ad7aed2b7540732

ongtw commented on requirements.txt:
```diff
@@ -11,4 +11,14 @@ tensorflow >= 2.2
+albumentations ~= 1.3
+matplotlib
+timm
+wandb
```
Todo: fix a version for each of the above packages.
Note: do not blindly copy and paste the version you see in your conda env. Think through what is a meaningful version to install, e.g. your conda may show `scikit-learn=1.2.2` but the "right" entry could be `scikit-learn >= 1.2.0`.

sabrimansor replied:
resolved

ongtw commented on peekingduck/training/test/__init__.py:
The pathname should be `peekingduck/training/tests`: plural, not singular.

sabrimansor replied:
Resolved in commit 73dc091721dc6aa35ab5ac014459e523127fb309

furyhawk replied:
peekingduck repository 2 modules contained within it. Each `tests` folder should have different names to differential the 2 modules `peekingduck` and `training pipeline`.  This is to hint `pytest` on how to import the correct module.

https://docs.pytest.org/en/7.1.x/explanation/goodpractices.html

ongtw commented on docs/source/training_pipeline/01_getting_started.rst:
```diff
@@ -0,0 +1,260 @@
+      .. admonition:: Terminal Session
+
+         | \ :blue:`[~user]` \ > \ :green:`git clone https://github.com/aisingapore/PeekingDuck.git` \
+         | \ :blue:`[~user]` \ > \ :green:`cd PeekdingDuck` \
```
Typo error: `peekingduck`.

Please run a spell checker (with custom dictionaries) on your docs.

frankcaoyun replied:
Updated.

ongtw commented on docs/source/training_pipeline/01_getting_started.rst:
```diff
@@ -0,0 +1,260 @@
+
+      .. admonition:: Terminal Session
+
+         | \ :blue:`[~user/PeekdingDuck]` \ > \ :green:`pip install -r peekingduck/training/requirements.txt` \
```
Typo error `peekingduck`.

There are many typo errors, which should not be there in the first place!

frankcaoyun replied:
The cause was that I accidentally added the typo word into my user dictionary. The typo word is now removed from the user dictionary and currently the spell checker is working as intended.

ongtw commented on docs/source/training_pipeline/01_getting_started.rst:
```diff
@@ -0,0 +1,260 @@
+View the results of each run at the specified output folder directory :mod:`\./PeekingDuck/outputs/\<PROJECT_NAME\>/\<DATE_TIME\>`, 
+\ where the default value of the :mod:`<PROJECT_NAME>` is defined in the :ref:`config-files-mainconfig`.
+
+After installation and test run, refer to :ref:`configuring_training_parameters` for training customizations.
```
Above has lots of errors: typos and grammar.
Also, pay attention to capitalisation and stylization: Weights & Biases, not weights and biases, or other format. Every small thing matter in the doc.
Please fix them, use spell checker, custom dictionaries, grammar checker, etc. All the tools are available.
I gave up halfway 'coz too tiring with so many errors, so don't just rely on my comments, read every single line again.

sabrimansor replied:
Resolved in commit edb09fa0c7a8f5e8d1f7e58add642ee16900f807

ongtw commented on docs/source/training_pipeline/index.rst:
```diff
@@ -0,0 +1,15 @@
+   /training_pipeline/01_getting_started
+   /training_pipeline/02_configuring_training_parameters/index
+   /training_pipeline/03_using_custom_dataset
+
```
Above organisation where `01` and `03` are files but `02` is a folder results in the following inconsistent view:
![image](https://user-images.githubusercontent.com/13075447/233833198-aa1b1959-9d35-4e44-bb4c-db3ed730810e.png)

Suggest to change the structure to the following:
```
01_getting_started.rst
02_configuring_training_parameters.rst
03_using_custom_dataset
training_parameters_configuration/   [this folder holds the original 02 files, referenced by second rst]
   data_module/
   trainer/
   use_case/
```

sabrimansor replied:
Noted and changed the naming and file structure

ongtw commented on peekingduck/training/README.md:
````diff
@@ -2,3 +2,32 @@
+Run on default cifar10 dataset:
+```sh
+python peekingduck/training/main.py
+```
````
If cifar10 dataset is removed from repo due to its size, can include instructions here to download it.

furyhawk replied:
`cifar10` dataset default to be downloaded from the cloud.

ongtw reviewed (changes requested):
Phase one comments.
To do:
1. Get macOS env setup script done
2. Fix documentation errors
3. Relink URLs to datasets in PeekingDuck GCP bucket

Then commit and will do phase 2 review.

sabrimansor committed:
refactor: renaming files to better fit the file structure

furyhawk committed:
feat: macos installation script, requirements.txt add version req

frankcaoyun committed:
docs: fix typo and formatting issues in the docs.

furyhawk committed:
fix: update url for cifar10

furyhawk committed:
feat: url for rsna and vegfru dataset

ongtw commented on peekingduck/training/scripts/install_tensorflow_macos.sh:
```diff
@@ -0,0 +1,123 @@
+    conda install -y bandit coverage mypy pylint pytest types-PyYAML types-requests \
+        types-setuptools types-six beautifulsoup4 myst-parser sphinx-autodoc-typehints \
+        texttable
+    pip install lap sphinx-rtd-theme
```
The above CI/CD packages are outdated for the new PKD training pipeline. Lots of missing packages, like `sphinx_copybutton`, etc. Please update and test this script by creating a brand new env, then doing a `make clean html` under `docs/` to verify that the install script is complete and correct.

furyhawk replied:
Fixed in commit https://github.com/furyhawk/PeekingDuck/commit/c94afa0f56080245ae5b73a24a6326899bff7a9f

ongtw commented on peekingduck/training/scripts/install_tensorflow_macos.sh:
```diff
@@ -0,0 +1,123 @@
+
+echo "-----------------------------"
+echo "Activate new environment with: conda activate $ENV_NAME"
+echo "Exiting $SELF"
```
Three issues:
1. Filename is wrong since PyTorch and Tensorflow are both installed.
2. Should put all scripts into `peekingduck/scripts` and not have two separate locations. The above should supersede the original macos install script.
3. When I manually fix the environment missing packages, and do a `make clean html` in `docs/`, I get the following errors:
```
WARNING: [autosummary] failed to import dabble.tracking.
Possible hints:
* AttributeError: module 'dabble' has no attribute 'tracking'
* AttributeError: module 'cv2' has no attribute 'legacy_TrackerMOSSE'
* ImportError:
WARNING: [autosummary] failed to import model.huggingface_hub.
Possible hints:
* ImportError:
* ModuleNotFoundError: No module named 'transformers'
* AttributeError: module 'model' has no attribute 'huggingface_hub'
WARNING: [autosummary] failed to import model.mediapipe_hub.
Possible hints:
* AttributeError: module 'model' has no attribute 'mediapipe_hub'
* ImportError:
* ModuleNotFoundError: No module named 'mediapipe'
```
Please see if you can fix the above.
4. It seems that `tensorflow-datasets` is missing from the installation? Because when I created a new env with this and run the unit tests, it gives error: `E   ModuleNotFoundError: No module named 'tensorflow_datasets'`

ongtw commented on peekingduck/training/requirements_mac.yaml:
```diff
@@ -0,0 +1,26 @@
+      - tqdm == 4.45.0
+      - albumentations >= 1.3.0
+      - matplotlib >= 3.6.3
+      - wandb >= 0.13.9
```
Is there a need for a separate `requirements_mac.yaml` when there is already the `requirements_mac.txt`?

furyhawk replied:
Removed `requirements_mac.yaml` and `requirements_mac.txt`, these two requirements are supercede by `install_training_pipeline_macos.sh`.  https://github.com/furyhawk/PeekingDuck/commit/32eb099171b8d243109e185e18df201502dd1023

ongtw commented on docs/source/training_pipeline/01_getting_started/index.rst:
```diff
@@ -0,0 +1,257 @@
+==============
+
+Training a custom model with custom datasets is one of the key tasks when building computer vision solutions.
+Peeking Duck allows you to train a custom computer vision model with your own dataset without the need to write the boilerplate codes.
```
The long lines (> 88 characters per line) problem still hasn't been fix.
Reformat this file to ensure each line wraps after 88 chars.
Did you `black` format this file? 'coz `black` will flag errors on long lines.
If not, then you need to `black` format this and other related files.

frankcaoyun replied:
Fixed in commit 7d0fba4fdd5e835dea6bba131fb3d6b0eb00cec0

ongtw commented on docs/source/training_pipeline/01_getting_started/index.rst:
```diff
@@ -0,0 +1,257 @@
+
+.. _setting-up-weights-and-biases:
+
+Setting up Weights & Biases
```
`up` -> `Up`
Properly stylized titles matter!

frankcaoyun replied:
Hi David, I searched a few guidelines and also took a reference from a [Python documentation](https://docs.python.org/3/using/index.html). It seems that according to the [APA Style](https://apastyle.apa.org/style-grammar-guidelines/capitalization/title-case) for title case, short prepositions (e.g., “as,” “at,” “by,” “for,” “in,” “of,” “off,” “on,” “per,” “to,” “up,” “via”) is recommended to use lowercase instead. Our code styles follows PEP8 but I'm not sure about the documentation style. Should we follow the APA Style for documentation?

frankcaoyun replied:
Keep the capitalized format.

ongtw commented on docs/source/training_pipeline/01_getting_started/index.rst:
```diff
@@ -0,0 +1,257 @@
+
+   <h6>Train with your own data</h6>
+
+PeekingDuck supports data loading and model training on your own dataset. Refer to :ref:`using_custom_dataset` for more details.
```
`Using Custom Dataset` -> stylize properly!

sabrimansor replied:
Resolved in commit 2469aa17c86058629ca4128f2308dade83328db6

ongtw commented on docs/source/training_pipeline/01_getting_started/index.rst:
```diff
@@ -0,0 +1,257 @@
+This section will introduce you the key features of the training pipeline and how to get started.
+The following sections will introduce you how to train your own model with your own data by customizing the training parameters.
+
+The current version of PeekingDuck supports image classification model training. Object detection are being developed and will be available in future releases.
```
`are` -> `is`.
Please proof-read your docs.

sabrimansor replied:
Resolved in commit 2469aa17c86058629ca4128f2308dade83328db6

ongtw commented on docs/source/training_pipeline/01_getting_started/index.rst:
```diff
@@ -0,0 +1,257 @@
+Peeking Duck allows you to train a custom computer vision model with your own dataset without the need to write the boilerplate codes.
+
+This section will introduce you the key features of the training pipeline and how to get started.
+The following sections will introduce you how to train your own model with your own data by customizing the training parameters.
```
Line 15 reads weird. I don't think we use such sentence structure in English. Did you run this through a grammar check?

sabrimansor replied:
Resolved

ongtw commented on docs/source/training_pipeline/02_configuring_training_parameters/index.rst:
```diff
@@ -0,0 +1,141 @@
+*******************************
+
+This section describes how you can configure the pipeline to suite your model training.
+\ The following parts will describe the image classification section and the documentation conventions used.
```
For this doc structure, propose putting `Documentation Convention` before `Image Classification` as one normally defines the convention first, then go into the technical details.

sabrimansor replied:
Resolved in commit 47ce3492bf4e6f810590fc19811cad907f1a8a79

ongtw commented on docs/source/training_pipeline/02_configuring_training_parameters/use_case/02a_image_classification.rst:
```diff
@@ -0,0 +1,1581 @@
+     <strong>Select Framework : </strong>
+     <button class="training-framework-btn training-framework-py active">Pytorch</button>
+     <button class="training-framework-btn training-framework-tf">Tensorflow</button>
+   </div>
```
Above `Select Framework :` is not working for me on macOS, using both Safari and Chrome browsers. Nothing happens when I click on the buttons: the highlight does not even switch. Same for all the other `Select Framework :` instances.

sabrimansor replied:
Resolved in commit dd3d2d3b125f544a3b3ac02ccc240b2305ae3e63

ongtw commented on docs/source/training_pipeline/02_configuring_training_parameters/use_case/02a_image_classification.rst:
```diff
@@ -0,0 +1,1581 @@
+Overview
+========
+
+We are using the yaml syntax for the config file. Below is the folder structure and description of how users can understand and navigate the config structure. 
```
`We are using` -> `We use`.

Also, 2nd sentence sounds grammatically weird.
`Below shows the folder structure and describes how users...` sounds more correct.
Again, run through grammer checker.

sabrimansor replied:
Resolved

ongtw commented on docs/source/training_pipeline/02_configuring_training_parameters/use_case/02a_image_classification.rst:
```diff
@@ -0,0 +1,1581 @@
+   :linenos: 
+
+   cd PeekingDuck
+   python ./peekingduck/training/main.py debug=True framework=tensorflow project_name=abcxyz view_only=True
```
`project_name=abcxyz` -> `project_name=my_project`.
Professional tech docs don't have random letters like `abcxyz` in places where you can meaningfully name something.
Please go through the rest of the docs, read in detail and catch all such instances.

sabrimansor replied:
Resolved

ongtw commented on docs/source/training_pipeline/03_using_custom_dataset/index.rst:
```diff
@@ -0,0 +1,190 @@
+.. _using_custom_dataset:
+
+********************
+Using custom dataset
```
`Using Custom Dataset` -> Stylize properly.

sabrimansor replied:
Resolved in commit 2469aa17c86058629ca4128f2308dade83328db6

ongtw commented on docs/source/training_pipeline/03_using_custom_dataset/index.rst:
```diff
@@ -0,0 +1,190 @@
+
+If you have already collected your own dataset, the following section describes\
+ how you can prepare your dataset, connect it to the training pipeline and use\
+ it for training a model.
```
The above line formatting looks weird, see below image:

![image](https://github.com/aisingapore/PeekingDuck/assets/13075447/970313b2-4f64-44ca-8a67-eb518baaa048)

sabrimansor replied:
Resolved in commit 2469aa17c86058629ca4128f2308dade83328db6

ongtw commented on docs/source/training_pipeline/03_using_custom_dataset/index.rst:
```diff
@@ -0,0 +1,190 @@
+
+.. raw:: html
+
+   <h6>Step 3. Edit config.yaml</h6>
```
Step 3 is formatted differently from Step 1 and Step 2 above.

sabrimansor replied:
Resolved in commit 2469aa17c86058629ca4128f2308dade83328db6

ongtw commented on docs/source/training_pipeline/03_using_custom_dataset/index.rst:
```diff
@@ -0,0 +1,190 @@
+Testing the Pipeline
+====================
+
+Refer to :ref:`getting-started-test-run` to test out the pipeline.
```
Instead of asking the reader to jump to another location, better to reproduce the Test Run codes here.

sabrimansor replied:
Resolved in commit 2469aa17c86058629ca4128f2308dade83328db6

ongtw commented on docs/source/training_pipeline/03_using_custom_dataset/index.rst:
```diff
@@ -0,0 +1,190 @@
+
+.. _custom_dataset:
+
+Label File Template
```
I think we can use a better name thatn `Label File Template`.
This part actually tells the user they need a `image`->`label` mapping file.
So title should be something like `Map Image to Label`.
And the file should be given a proper name, like `image_to_label_map.csv`

sabrimansor replied:
Resolved in commit @sabrimansor  2469aa17c86058629ca4128f2308dade83328db6

ongtw commented on docs/source/training_pipeline/03_using_custom_dataset/index.rst:
```diff
@@ -0,0 +1,190 @@
+            │             ├── <image_002>.jpg
+            │             ├── <image_003>.jpg
+            │             └── ...
+            └── <your_csv_file>.yaml
```
The file should not be called `your_csv_file.yaml`: is it a CSV or is it a YAML file??
This confuses the reader.
Is it `image_to_label_map.csv` or `image_to_label_map.yaml`?

sabrimansor replied:
Resolved in commit 2469aa17c86058629ca4128f2308dade83328db6

ongtw commented on docs/source/training_pipeline/03_using_custom_dataset/index.rst:
```diff
@@ -0,0 +1,190 @@
+            │      └── 1. <dataset_filename>.yaml
+            ├── \ :blue:`resample/...` \ |Blank|
+            ├── \ :blue:`transform/...` \ |Blank|
+            └── 2. <dataset_filename>.yaml   
```
Why are two `<dataset_filename>.yaml` needed when they are the same file?
If they are the same, then there should only be one file. The "other file" can just hardcode to point to the first.
If they are different, then they should not have the same names.

sabrimansor replied:
Change added in commit:  2469aa17c86058629ca4128f2308dade83328db6

ongtw commented on docs/source/training_pipeline/03_using_custom_dataset/index.rst:
```diff
@@ -0,0 +1,190 @@
+| data/vegfru/veg200_images/Chinese_artichoke/v_09_03_0031.jpg | 8        | Chinese_pumpkin   |
++--------------------------------------------------------------+----------+-------------------+
+
+Preparation & Integrating with Training Pipeline
```
Use `and` instead of `&`.
The title doesn't quite reflect the following sections: the `integration` part is not obvious, looks like all `prep`.
But see all the way below for a major restructure proposal of this whole doc.

sabrimansor replied:
Resolved in commit 2469aa17c86058629ca4128f2308dade83328db6

ongtw commented on docs/source/training_pipeline/03_using_custom_dataset/index.rst:
```diff
@@ -0,0 +1,190 @@
+
+Refer to :ref:`getting-started-test-run` to test out the pipeline.
+
+This is the end of the documentation for the training pipeline. The next section will describe the PeekingDuck ecosystem.
```
This `Using Custom Dataset` doc feels a bit confusing.
It'll be better to structure it as follows:
```
* Intro
* Setup
   1 Image to Label Mapping
   2 Data Folder Structure
   3 Configuration Files
      a) dataset_filename.yaml
      b) dataset_filename.yaml   (now you see the confusion?!)
      c) config.yaml
* Run
```

In fact, having a diagram in introduction to show how the different files connect and flow will probably make things a lot clearer to the reader.

sabrimansor replied:
Resolved in commit 2469aa17c86058629ca4128f2308dade83328db6

ongtw commented on peekingduck/training/configs/metrics/detection.yaml:
If empty file, then should remove it to avoid confusing the user.

furyhawk replied:
Resolved in commit https://github.com/furyhawk/PeekingDuck/commit/32eb099171b8d243109e185e18df201502dd1023

ongtw commented on peekingduck/training/configs/model/detection.yaml:
Empty file should be deleted.

furyhawk replied:
Resolved in commit https://github.com/furyhawk/PeekingDuck/commit/32eb099171b8d243109e185e18df201502dd1023

ongtw commented on peekingduck/training/configs/trainer/detection.yaml:
Remove empty file.

furyhawk replied:
Resolved in commit https://github.com/furyhawk/PeekingDuck/commit/32eb099171b8d243109e185e18df201502dd1023

ongtw commented on peekingduck/training/configs/use_case/classification.yaml:
Remove empty file

furyhawk replied:
Use as placeholder for classification use case `Hydra` interpolation.

Explanation comment added in commit https://github.com/furyhawk/PeekingDuck/commit/32eb099171b8d243109e185e18df201502dd1023

ongtw commented on peekingduck/training/configs/use_case/cifar10.yaml:
Remove empty file?

furyhawk replied:
Resolved in commit https://github.com/furyhawk/PeekingDuck/commit/32eb099171b8d243109e185e18df201502dd1023

ongtw commented on peekingduck/training/configs/use_case/detection.yaml:
Remove empty file

furyhawk replied:
Resolved in commit https://github.com/furyhawk/PeekingDuck/commit/32eb099171b8d243109e185e18df201502dd1023

ongtw commented on peekingduck/training/test/test_training/tensorflow_training.py:
```diff
@@ -0,0 +1,60 @@
+
+probability_model = tf.keras.Sequential([model, tf.keras.layers.Softmax()])
+
+probability_model(x_test[:5])
```
Question: If this test runs successfully, it only proves that the Tensorflow training can run from start to end. But where is the test of correctness? There should be some kind of `assert(trained_model_inf_result == expected_inf_result)` towards the end to test correctness. Or am I missing something?

furyhawk replied:
`pytorch_training.py` and `tensorflow_training.py` without the `test_` prefix are used for internal testing on new machine. For example, to test whether the cuda library is installed correctly.

However, these to scripts are under `test` folder and pytest wouldn't run them without the `test_` prefix. May lead to some confusion on their purpose.

ongtw reviewed (changes requested):
The Good: Managed to create env on macOS and run the unit tests. 👍🏼
Still to improve: Documentation and some stuff. Please review all the comments.

furyhawk committed:
fix: CICD requirements to macos install script

furyhawk committed:
fix: remove h5py downgrade

furyhawk committed:
refactor: remove detection configurations

sabrimansor committed:
Refactor: changed title casing and added jquery library extension

furyhawk committed:
test: add  tensorflow-datasets to CICD, print out loss value

furyhawk committed:
fix: remove extra comments

frankcaoyun committed:
doc: reduce line length in getting started file

frankcaoyun committed:
Merge branch 'feat-training' of https://github.com/furyhawk/PeekingDuck into feat-training

sabrimansor committed:
docs: fixed some phrasing based on comments

sabrimansor reviewed (commented):
refactored main config parameters to resolve issue of duplicate files in data_module folder

furyhawk committed:
fix: set seed_all for reproducibility

sabrimansor committed:
refactor: change data module config structure to remove files with the same name

sabrimansor committed:
fix: black reformat error

sabrimansor committed:
docs: added change from comments

sabrimansor committed:
refactor: change folder name to plural instead

sabrimansor committed:
docs: added fix for typo errors

sabrimansor committed:
docs: reorganise sections

frankcaoyun committed:
doc: Update the write-up for the introduction. Fix typos and line length issues.

furyhawk committed:
fix: refactor to "data_module/dataset=cifar10"

frankcaoyun committed:
Merge branch 'feat-training' of https://github.com/furyhawk/PeekingDuck into feat-training

frankcaoyun committed:
Merge branch 'feat-training' of https://github.com/furyhawk/PeekingDuck into feat-training

frankcaoyun committed:
doc: Correction for formatting issues.

sabrimansor committed:
docs: converting html tables to sphinx syntax for better maintainability part 1

furyhawk committed:
fix: remove __init__.py from tests folders

furyhawk committed:
fix: missing init.py in training/tests/

furyhawk committed:
refactor: tests to test folder revert

sabrimansor committed:
docs: converted html tables to sphinx syntax part 2

furyhawk committed:
refactor: rename tests to tests_training_pipeline

furyhawk committed:
Merge branch 'feat-training' of https://github.com/furyhawk/PeekingDuck into feat-training

sabrimansor committed:
docs: converting html table to sphinx syntax part 3

sabrimansor committed:
Merge remote-tracking branch 'refs/remotes/origin/feat-training' into feat-training

sabrimansor committed:
docs: fixed typos, grammar errors and dead links

furyhawk committed:
refactor: rename install_macos.sh to training pipeline use case

furyhawk committed:
refactor: remove mac requirements.txt as it is supercede by mac install script

sabrimansor committed:
docs: correct grammar and spelling issues

frankcaoyun committed:
docs: fix formatting issues for headers.

sabrimansor committed:
docs: added fix for grammar and spelling errors

sabrimansor committed:
docs: added fix for grammar and misspelling

frankcaoyun committed:
docs: improve descriptions write-ups. Fix line length issues.

sabrimansor committed:
docs: added line breaks to description

frankcaoyun committed:
doc: fix issues with doc writing and formatting.

frankcaoyun committed:
docs: fix installation guide for mac with the updated script.

frankcaoyun committed:
docs: fix formatting issue for terminal commands.

sabrimansor committed:
docs: edit files and folder directory tree to reflect changes

frankcaoyun committed:
docs: simplify and clarify the data folder structure.

frankcaoyun committed:
Merge branch 'feat-training' of https://github.com/furyhawk/PeekingDuck into feat-training

furyhawk committed:
fix: group `tensorflow-datasets` to pip tensorflow

found that it's a issue with `tensorflow-datasets` was installed using `conda` in the CICD portion. This will install an older version of `tensorflow-estimator=2.11`, while ```pip install tensorflow tensorflow-metal``` install `tensorflow-estimator=2.13-rc1`.  Fixed and pytested on python 3.8, 3.10 by grouping ```pip install tensorflow tensorflow-metal tensorflow-datasets```.

furyhawk committed:
refactor: decouple project_name from dataset folder

furyhawk committed:
doc: update project_name usage. add dataset var for dataset folder

frankcaoyun committed:
docs: Clarify the descriptions after decoupling <project_name>.

ongtw commented on peekingduck/training/tests_training_pipeline/test_training/test_training_pipeline.py:
```diff
@@ -0,0 +1,133 @@
+from typing import List
+
+import sys
+from pytest import mark
```
To group above 3 system imports into one block.

furyhawk replied:
https://github.com/furyhawk/PeekingDuck/tree/47df8145fbba97081c6758cf66856b571d37db99

Grouped.

ongtw commented on peekingduck/training/tests_training_pipeline/test_training/tensorflow_training.py:
```diff
@@ -0,0 +1,60 @@
+
+probability_model = tf.keras.Sequential([model, tf.keras.layers.Softmax()])
+
+probability_model(x_test[:5])
```
Likewise, need assert test condition here.

furyhawk replied:
https://github.com/furyhawk/PeekingDuck/tree/47df8145fbba97081c6758cf66856b571d37db99

Removed for internal testing file.

ongtw commented on peekingduck/training/tests_training_pipeline/test_training/pytorch_training.py:
```diff
@@ -0,0 +1,83 @@
+
+        running_loss += loss.item()
+
+    print(f"Epoch {epoch+1}: loss={running_loss/len(data_loader)}")
```
Why is there no assert test condition in above code?
We should check some basic model accuracy numbers before and after training loop, to ensure that "it works".

furyhawk replied:
https://github.com/furyhawk/PeekingDuck/tree/47df8145fbba97081c6758cf66856b571d37db99

Removed for internal testing file.

ongtw commented on peekingduck/training/src/trainer/tensorflow_trainer.py:
```diff
@@ -0,0 +1,177 @@
+from omegaconf import DictConfig
+import tensorflow as tf
+from tqdm.keras import TqdmCallback  # for progress bar
+from configs import LOGGER_NAME
```
To separate the above imports into two blocks: Python system imports and third party imports.

furyhawk replied:
https://github.com/furyhawk/PeekingDuck/tree/47df8145fbba97081c6758cf66856b571d37db99

Grouped.

ongtw commented on peekingduck/training/src/training_pipeline.py:
```diff
@@ -0,0 +1,90 @@
+from time import perf_counter
+from omegaconf import DictConfig
+from hydra.utils import instantiate
+from configs import LOGGER_NAME
```
To separate the above into two blocks: Python imports and third party imports

furyhawk replied:
https://github.com/furyhawk/PeekingDuck/tree/47df8145fbba97081c6758cf66856b571d37db99

Grouped.

ongtw commented on peekingduck/training/src/optimizers/schedules.py:
```diff
@@ -0,0 +1,69 @@
+    def CosineDecayRestarts(parameters):
+        return tf.keras.optimizers.schedules.CosineDecayRestarts(**parameters)
+
+"""
```
Is the above comment supposed to help the developer use the different `Keras` optimizers? If yes, pls explicitly mention it in the comments (i.e. how to use). If not, then they should be removed.

furyhawk replied:
https://github.com/furyhawk/PeekingDuck/tree/47df8145fbba97081c6758cf66856b571d37db99

Removed comment.

ongtw commented on peekingduck/training/src/optimizers/adapter.py:
```diff
@@ -0,0 +1,103 @@
+        return tf.keras.optimizers.SGD(learning_rate=learning_rate, **parameters)
+
+
+"""
```
Is above comment supposed to help user use the `Keras` optimizers? If yes, it should be state how to use. Otherwise, it should be removed.

furyhawk replied:
https://github.com/furyhawk/PeekingDuck/tree/47df8145fbba97081c6758cf66856b571d37db99

Removed comment.

ongtw reviewed (changes requested):
Some code comments for your consideration.

Still going through the .rst docs. Will get back on those.

furyhawk committed:
refactor: cleanup of comments and extra files. import grouping

ongtw reviewed (approved):
Ok
